### PR TITLE
Adding validation rules for unique resource paths

### PR DESCRIPTION
**Pull request by dsgouda**
- Contains validation rules for Non app json media types and Non-Https schemes 
- Fixing documentation to account for changes to clean json when adding new validation rules
- Adding validation rule that ensures we reference a unique resource path within the given spec
Addressing issues https://github.com/Azure/autorest/issues/1499 , #1516 and #1517
@tbombach PTAL, also need you to verify the error message here

### CODE REVIEW

File: src/modeler/AutoRest.Swagger.Tests/Swagger/Validation/positive/clean-complex-spec.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
   "host": "petstore.swagger.io",
   "basePath": "/v1",
   "schemes": [
-    "http"
+    "Https"
```
Comment (fearthecowboy) [inline src/modeler/AutoRest.Swagger.Tests/Swagger/Validation/positive/clean-complex-spec.json]: Case? shouldn't this be `https` ?

And do we not support `http`?

File: src/core/AutoRest.Core/Properties/Resources.resx [hunk 2]
```diff
@@ -252,4 +252,13 @@
   <data name="InvalidConstraint" xml:space="preserve">
     <value>Constraint is not supported for this type and will be ignored.</value>
   </data>
+  <data name="NonAppJsonTypeNotSupported" xml:space="preserve">
+    <value>Non application json MIME type has limited support</value>
```
Comment (fearthecowboy) [inline src/core/AutoRest.Core/Properties/Resources.resx]: How about `Support for media types other than "application/json" has limited support` 

Points to note: 
- be explicit about what someone should use (`"application/json"` )
- `MIME type` is actually not the correct term anymore . It's `Media Type` (it not being used to describe Multipurpose Internet Mail Extensions anymore...)

File: src/core/AutoRest.Core/Properties/Resources.resx [hunk 3]
```diff
@@ -252,4 +252,13 @@
...
+  </data>
+  <data name="SupportedSchemesWarningMessage" xml:space="preserve">
+    <value>Non-HTTPS schemes have limited Autorest support</value>
```
Comment (fearthecowboy) [inline src/core/AutoRest.Core/Properties/Resources.resx]: o_O what are we saying about HTTP? We support HTTP, right?

Again, maybe be clearer `Support for schemes other than {"http","https" } have limited support `

File: src/core/AutoRest.Core/Properties/Resources.resx [hunk 4]
```diff
@@ -252,4 +252,13 @@
...
+  </data>
+  <data name="NetworkInterfacesApiWarningMessage" xml:space="preserve">
+    <value>Swagger API can reference either Microsoft.Compute or Microsoft.Network Apis and not both</value>
```
Comment (fearthecowboy) [inline src/core/AutoRest.Core/Properties/Resources.resx]: `Swagger` is deprecated. The new term is `OpenAPI` ... but maybe we should drop the reference to the format, and say something like `May not reference more than one of {"Microsoft.Compute","Microsoft.Network"}` 

- this makes it so we can add another item to the list without substantially changing the message later.

File: src/modeler/AutoRest.Swagger/Validation/NetworkInterfacesApiWarning.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        private const string MicrosoftNetworkApi = "Microsoft.Network";
+        /// <summary>
+        /// This rule passes if the paths contain reference to either Microsoft.Network Apis or 
+        /// Microsoft.Compute Apis but not both
+        /// </summary>
+        /// <param name="paths"></param>
+        /// <returns></returns>
+        public override bool IsValid(Dictionary<string, Dictionary<string, Operation>> paths)
+        {
+            Regex ApiRegExp = new Regex(@"(/.+)+/((M|m)icrosoft.(C|c)ompute|(M|m)icrosoft.(N|n)etwork)/.+");
```
Comment (fearthecowboy) [inline src/modeler/AutoRest.Swagger/Validation/NetworkInterfacesApiWarning.cs]: `MICROSOFT.COMPUTE` will pass thru that. 

Perhaps we want a case insensitive regex?

File: src/core/AutoRest.Core/Properties/Resources.resx (hunk 3 above)
Comment (dsgouda) [inline src/core/AutoRest.Core/Properties/Resources.resx]: I believe @tbombach indicated we support https only (not http)

File: src/modeler/AutoRest.Swagger/Validation/NonAppJsonTypeWarning.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    public class NonAppJsonTypeWarning : TypedRule<string>
+    {
+        private const string AppJsonType = "application/json";
+        /// <summary>
+        /// This rule passes if the entity contains application/json type 
+        /// </summary>
+        /// <param name="entity"></param>
+        /// <returns></returns>
+        public override bool IsValid(string entity)
+            => (entity.ToLowerInvariant().StartsWith(AppJsonType));
```
Comment (fearthecowboy) [inline src/modeler/AutoRest.Swagger/Validation/NonAppJsonTypeWarning.cs]: `StartsWith` ? 

Hmm. that's probably a good idea, good catch. I can't see a case where we shouldn't do that, but there may be one day.

**Comment by fearthecowboy** [comment]
LGTM 👍

**Comment by tbombach** [comment]
(As mentioned offline, Deepak is going to make this rule more generic. The rule will check that paths only contain one RP path, not for Network/Compute specifically)

**Comment by tbombach** [comment]
Minor: make sure the PR title is updated, since the rule is more generic now

**Comment by tbombach** [comment]
This rule should work for all Azure services, without having to add a resource type to the list to check. The general logic will be:

1. Create an empty list of resource types names.
2. Iterate through each path
3. Capture the RP resource type with a regex (i.e., given a path "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/microsoft.Foo", it will capture "Foo") and add it to the list of resource type names
  - Make sure that there's no exception thrown or problem if the path is not of the above form. That just means we should skip that path, because it won't affect whether or not this rule passes.
4. Make sure that the list of resource types is unique

You could optimize this by only storing the first resource type name that you find and failing the rule as soon as you see a different one.

**Comment by fearthecowboy** [comment]
LGTM 👍

:shipit:

**Comment by dsgouda** [comment]
@azuresdkci retest this please

File: src/modeler/AutoRest.Swagger/Validation/UniqueResourcePaths.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+        /// <summary>
+        /// This rule passes if the paths contain reference to exactly one of the namespace resources
+        /// </summary>
+        /// <param name="paths"></param>
+        /// <returns></returns>
+        public override bool IsValid(Dictionary<string, Dictionary<string, Operation>> paths) 
+            => paths.Keys.Select(path => 
+                    { return apiRegExp.Match(path).Success ? apiRegExp.Match(path).Groups?[0]?.Value.ToString().ToLowerInvariant() : null; })
+                                      .Except(new List<string> { null }).Distinct().Count() <= 1;
```
Comment (tbombach) [inline src/modeler/AutoRest.Swagger/Validation/UniqueResourcePaths.cs]: I think `.Where(p => p != null)` is preferable over `.Except` for readability, plus it doesn't instantiate a new List to filter nulls

**Comment by tbombach** [comment]
Only things to change is the `.Except` to `.Where`, then it looks good to me (feel free to merge after updating the PR)

**Pull request merged**